### PR TITLE
added examples for ISO mDL 18013-5 data model

### PR DESCRIPTION
**Pull request by Sakurann**
Addresses Issue #144 and expresses ISO 18013-5 mDL data model using SD-JWT.
18013-5 itself is in CBOR, so this SD-JWT includes all of the data elements/use claims that are mandatory by 18013-5 in JSON.

uses same data set as in [OID4VP ISO mDL example](https://openid.net/specs/openid-4-verifiable-presentations-1_0.html#name-iso-mobile-driving-licence-)

Given EUDIW ARF, this PR and PR #210 should probably be combined into one.

### CODE REVIEW

File: draft-ietf-oauth-selective-disclosure-jwt.md [hunk 1]
```diff
@@ -1068,6 +1078,42 @@ Disclosures:
 
 {{examples/w3c-vc/disclosures.md}}
 
+## Example 5 - ISO/IEC 18013-5 mDL
+
+This example illustrates how the artifacts defined in this specification can
+be used to express an mDL data model defined in ISO/IEC 18013-5 [@ISO.18013-5] encoded as JSON.
+
+SD-JWT is equivalent to an `issuerSigned` object as defined in [@ISO.18013-5].
```
Comment (danielfett) [inline draft-ietf-oauth-selective-disclosure-jwt.md:1086]:
```suggestion
An SD-JWT is equivalent to an `issuerSigned` object as defined in [@ISO.18013-5].
```

File: draft-ietf-oauth-selective-disclosure-jwt.md [hunk 2]
```diff
@@ -1068,6 +1078,42 @@ Disclosures:
...
+
+The payload of a corresponding SD-JWT looks as follows:
+
+<{{examples/18013-5/sd_jwt_payload.json}}
+
+Disclosures:
+
+{{examples/18013-5/disclosures.md}}
+
+A Combined Format for Presentation for the SD-JWT that discloses only discloses `nameSpace`, `issuing_authority` and `issuing_country`, and has a Holder Binding JWT could look as follows:
```
Comment (danielfett) [inline draft-ietf-oauth-selective-disclosure-jwt.md:1109]:
```suggestion
A Combined Format for Presentation for the SD-JWT that discloses only `nameSpace`, `issuing_authority`, and `issuing_country`, and has a Holder Binding JWT, could look as follows:
```

File: sd_jwt/examples/18013_5.yml [hunk 3]
```diff
@@ -0,0 +1,46 @@
+only copy-paste, needs to be updated:
```
Comment (danielfett) [inline sd_jwt/examples/18013_5.yml:1]: Is this intended to be here?

**Review by danielfett — APPROVED** [review]

**Comment by danielfett** [comment]
I can't comment on the validity of the mDL-specifics, but overall this looks good.

File: sd_jwt/examples/18013_5.yml (hunk 3 above)
Comment (peppelinux) [reply sd_jwt/examples/18013_5.yml:1]: the same by me, shouldn't this text be removed?

**Review by peppelinux — CHANGES_REQUESTED** [review]

File: sd_jwt/examples/18013_5.yml (hunk 3 above)
Comment (Sakurann) [inline sd_jwt/examples/18013_5.yml:1]:
```suggestion
```

File: sd_jwt/examples/18013_5.yml [hunk 4]
```diff
@@ -0,0 +1,46 @@
+only copy-paste, needs to be updated:
+user_claims:
+  iss: "https://example.com"
+  cnf:
+    jwk:
+      kty: RSA
+      "n": 0vx7agoebGcQSu....-csFCur-kEgU8awapJzKnqDKgw
```
[the remark this answers is not in the document]
Comment (Sakurann) [reply sd_jwt/examples/18013_5.yml:7]: I had the same question, but I think it works even without ". (@danielfett)

**Comment by Sakurann** [comment]
the text needs to be updated to the new terminology and the yml file needs to be reformatted, but after that should be good to go.

File: draft-ietf-oauth-selective-disclosure-jwt.md [hunk 5]
```diff
@@ -1068,6 +1078,42 @@ Disclosures:
...
 {{examples/w3c-vc/disclosures.md}}
 
+## Example 5 - ISO/IEC 18013-5 mDL
+
+This example illustrates how the artifacts defined in this specification can
+be used to express an mDL data model defined in ISO/IEC 18013-5 [@ISO.18013-5] encoded as JSON.
+
+An SD-JWT is equivalent to an `issuerSigned` object as defined in [@ISO.18013-5].
+
+A Combined Format for Presentation with a Holder Binding JWT is equivalent to a Document
```
Comment (bc-pi) [inline draft-ietf-oauth-selective-disclosure-jwt.md:1088]:
```suggestion
An SD-JWT Presentation with a Key Binding JWT is equivalent to a Document
```

File: sd_jwt/examples/18013_5.yml [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+  {
+    "nameSpaces":
+      {
+        "nameSpace": null,
+        "issuing_authority": null,
+        "issuing_country": null
+      },
+  }
+
+holder_binding: True
```
Comment (bc-pi) [inline sd_jwt/examples/18013_5.yml:45]:
```suggestion
key_binding: True
```

File: sd_jwt/examples/18013_5.yml [hunk 7]
```diff
@@ -0,0 +1,45 @@
+user_claims:
+  iss: "https://example.com"
+  cnf:
```
Comment (bc-pi) [inline sd_jwt/examples/18013_5.yml:3]: Is there reason to not let it use the holder_key from settings.yml? I think that's what will sign the KB-JWT anyway...

File: examples/18013-5/verified_contents.json [hunk 8]
```diff
@@ -0,0 +1,26 @@
+{
```
Comment (bc-pi) [inline examples/18013-5/verified_contents.json:1]: don't need this file anymore

File: examples/18013-5/user_claims.json [hunk 9]
```diff
@@ -0,0 +1,65 @@
+{
```
Comment (bc-pi) [inline examples/18013-5/user_claims.json:1]: don't need this file anymore

File: examples/18013-5/sd_jwt_serialized.txt [hunk 10]
```diff
@@ -0,0 +1,23 @@
+eyJhbGciOiAiRVMyNTYifQ.eyJpc3MiOiAiaHR0cHM6Ly9leGFtcGxlLmNvbSIsICJpY
```
Comment (bc-pi) [inline examples/18013-5/sd_jwt_serialized.txt:1]: don't need this file anymore

File: examples/18013-5/sd_jwt_payload.json [hunk 11]
```diff
@@ -0,0 +1,38 @@
+{
```
Comment (bc-pi) [inline examples/18013-5/sd_jwt_payload.json:1]: don't need this file anymore

File: examples/18013-5/hb_jwt_serialized.txt [hunk 12]
```diff
@@ -0,0 +1,35 @@
+eyJhbGciOiAiRVMyNTYifQ.eyJub25jZSI6ICIxMDhkYTk5MDhjMDMyMDRiYmExMDdiM
+WM4OTZjYmJiZiIsICJhdWQiOiAiaHR0cHM6Ly9leGFtcGxlLmNvbS92ZXJpZmllciIsI
+CJpYXQiOiAxNjc4NDk4NjcwfQ.VDkAPf7XlT2X0pJEicWcRcLXAgBZAy5kEcHTlsPTGh
+MDV664TwPBT_Jo9EsZv_gFvAyFfJkO0u03PKdm_GD2tw
+Combined representation of SD-JWT and HS-Disclosures (combined_presentation):
```
Comment (bc-pi) [inline examples/18013-5/hb_jwt_serialized.txt:5]: huh? but whatever - don't need this file anymore

File: examples/18013-5/hb_jwt_payload.json [hunk 13]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (bc-pi) [inline examples/18013-5/hb_jwt_payload.json:1]: don't need this file anymore

File: examples/18013-5/disclosures.md [hunk 14]
```diff
@@ -0,0 +1,270 @@
+__Disclosure for `nameSpace`:__
```
Comment (bc-pi) [inline examples/18013-5/disclosures.md:1]: don't need this file anymore

File: examples/18013-5/combined_presentation.txt [hunk 15]
```diff
@@ -0,0 +1,30 @@
+eyJhbGciOiAiRVMyNTYifQ.eyJpc3MiOiAiaHR0cHM6Ly9leGFtcGxlLmNvbSIsICJpY
```
Comment (bc-pi) [inline examples/18013-5/combined_presentation.txt:1]: don't need this file anymore

File: examples/18013-5/combined_issuance.txt [hunk 16]
```diff
@@ -0,0 +1,67 @@
+eyJhbGciOiAiRVMyNTYifQ.eyJpc3MiOiAiaHR0cHM6Ly9leGFtcGxlLmNvbSIsICJpY
```
Comment (bc-pi) [inline examples/18013-5/combined_issuance.txt:1]: don't need this file anymore

File: draft-ietf-oauth-selective-disclosure-jwt.md [hunk 17]
```diff
@@ -1068,6 +1078,42 @@ Disclosures:
...
+
+The payload of a corresponding SD-JWT looks as follows:
+
+<{{examples/18013-5/sd_jwt_payload.json}}
+
+Disclosures:
+
+{{examples/18013-5/disclosures.md}}
+
+A Combined Format for Presentation for the SD-JWT that discloses only `nameSpace`, `issuing_authority`, and `issuing_country`, and has a Holder Binding JWT, could look as follows:
```
Comment (bc-pi) [inline draft-ietf-oauth-selective-disclosure-jwt.md:1109]:
```suggestion
An SD-JWT Presentation that discloses only `nameSpace`, `issuing_authority`, and `issuing_country`, and has a Key Binding JWT, could look as follows:
```

File: draft-ietf-oauth-selective-disclosure-jwt.md [hunk 18]
```diff
@@ -1068,6 +1078,42 @@ Disclosures:
...
+be used to express an mDL data model defined in ISO/IEC 18013-5 [@ISO.18013-5] encoded as JSON.
+
+An SD-JWT is equivalent to an `issuerSigned` object as defined in [@ISO.18013-5].
+
+A Combined Format for Presentation with a Holder Binding JWT is equivalent to a Document
+structure that contains `docType`, `issuerSigned` object and `deviceSigned` object
+as defined in [@ISO.18013-5].
+
+In this example, Holder Binding is applied and a Combined Format for Presentation is signed
+using a raw public key passed in a `cnf` Claim in the SD-JWT.
```
Comment (bc-pi) [inline draft-ietf-oauth-selective-disclosure-jwt.md:1093]:
```suggestion
In this example, Key Binding is applied and the Key Binding JWT is signed with the private key corresponding to
the raw public key passed in a `cnf` Claim in the SD-JWT.
```

File: draft-ietf-oauth-selective-disclosure-jwt.md [hunk 19]
```diff
@@ -1068,6 +1078,42 @@ Disclosures:
...
+
+<{{examples/18013-5/sd_jwt_payload.json}}
+
+Disclosures:
+
+{{examples/18013-5/disclosures.md}}
+
+A Combined Format for Presentation for the SD-JWT that discloses only `nameSpace`, `issuing_authority`, and `issuing_country`, and has a Holder Binding JWT, could look as follows:
+
+<{{examples/18013-5/combined_presentation.txt}}
```
Comment (bc-pi) [inline draft-ietf-oauth-selective-disclosure-jwt.md:1111]:
```suggestion
<{{examples/18013-5/sd_jwt_presentation.txt}}
```

**Review by bc-pi — CHANGES_REQUESTED** [review]
Lots of little changes needed to bring this PR up to date with recent(ish) changes to the way example content is generated and terminology in document itself.

File: examples/18013-5/sd_jwt_payload.json [hunk 20]
```diff
@@ -0,0 +1,38 @@
...
+  "exp": 1678517570,
+  "cnf": {
+    "jwk": {
+      "kty": "EC",
+      "crv": "P-256",
+      "x": "7jIrJ0veNwJa4p7WoDl9stzlxvgj3X4autz8NkR7vWQ",
+      "y": "cDE-9B9gO6h-OSA033Ua3wfVtHEdl-WBHC3dxNUkoAY"
+    }
+  },
+  "docType": "org.iso.18013.5.1.mDL",
```
Comment (peppelinux) [inline examples/18013-5/sd_jwt_payload.json:13]: ok, let's try do to some clarification.

this representation doesn't contain the elements of a typical mdoc structure like the one below

````
{
 'version': '1.0',
 'documents': Array of $doc,
 'status': 0
}
````
but just a doc, and its ok only if we explicitly say it in the text. At the same time we find the claim `version` in the proposed example, so it seems a mix. Just give some clarification about this not conventional data structure.

Going to the single doc (considering that an mdoc is able to contain more than a single doc), we have

````
{
 "docType": "org.iso.18013.5.1.mDL",
 "issuerSigned": {
     "nameSpaces":  {
         "org.iso.18013.5.1": [ $array of digests ]}, 
     "issuerAuth": {CBORTag 18 containing a COSE Sign1 Document without a content-type, that's the MSO}
 },
 "deviceSigned": ... not of interest for SD-JWT
}
````

I agree about the flattened approach proposed in this PR, that makes sd-jwt one step forward.

Probably the nameSpaces may have the `nameSpace` as a dictionary, as it is in mdoc cbor

**Review by peppelinux — CHANGES_REQUESTED** [review]

**Comment by Sakurann** [comment]
closing in favor of #304. the work could continue in sd-jwt-vc repo

**Pull request closed without merging**